common/guestchain/src/config.rs

```diff
@@ -91,3 +91,104 @@ pub struct Config {
     /// epoch.
     pub min_epoch_length: crate::height::HostDelta,
 }
+
+#[derive(Clone, Debug, borsh::BorshSerialize, borsh::BorshDeserialize)]
+pub struct UpdateConfig {
+    pub min_validators: Option<NonZeroU16>,
+    pub max_validators: Option<NonZeroU16>,
+    pub min_validator_stake: Option<NonZeroU128>,
+    pub min_total_stake: Option<NonZeroU128>,
+    pub min_quorum_stake: Option<NonZeroU128>,
+    pub min_block_length: Option<crate::height::HostDelta>,
+    pub max_block_age_ns: Option<u64>,
+    pub min_epoch_length: Option<crate::height::HostDelta>,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum UpdateConfigError {
+    /// Minimum validators are more than existing
+    ///
+    /// If minimum validators are higher than existing, then the
+    /// none of the existing validators can leave unless the validators are more
+    /// than the minimum.
+    MinValidatorsHigherThanExisting,
+
+    /// Maximum validators should always be equal or higher than minimum validators
+    MaxValidatorsCannotBeLowerThanMin,
+
+    /// Minimum Total Stake is higher than existing
+    ///
+    /// If minimum total stake is higher than existing, then none of them
+    /// can withdraw their unless the total stake is more than the minimum.
+    MinTotalStakeHigherThanExisting,
+
+    /// Minimum total stake is lower than minimum quorum stake
+    ///
+    /// If minimum total stake is lower than minimum quorum stake, then
+    /// the total stake can be less than quorum and block would never be
+    /// finalized since the total stake is less than quroum stake.
+    MinTotalStakeHigherThanMinQuorumStake,
+
+    /// Minimum Quorum Stake is higher than existing total stake
+    ///
+    /// If minimum quorum stake is higher than existing total stake, then
+    /// blocks would never get finalized until more stake is added and quorum
+    /// stake is less than head stake.
+    MinQuorumStakeHigherThanTotalStake,
+}
+
+impl Config {
+    pub fn update(
+        &mut self,
+        head_stake: u128,
+        total_validators: u16,
+        config_payload: UpdateConfig,
+    ) -> Result<(), UpdateConfigError> {
+        if let Some(min_validators) = config_payload.min_validators {
+            if min_validators > NonZeroU16::new(total_validators).unwrap() {
+                return Err(UpdateConfigError::MinValidatorsHigherThanExisting);
+            }
+            self.min_validators = min_validators;
+        }
+        if let Some(max_validators) = config_payload.max_validators {
+            if max_validators < self.min_validators {
+                return Err(
+                    UpdateConfigError::MaxValidatorsCannotBeLowerThanMin,
+                );
+            }
```

Comment:
well we just need to make sure that new minimum is less than the new maximum right? so we could prolly have smtg like this inside `min_validators` condn
```rs
if let Some(min_validators) = config_payload.min_validators {
            if min_validators > NonZeroU16::new(total_validators).unwrap() {
                return Err(UpdateConfigError::MinValidatorsHigherThanExisting);
            }
            if let Some(max_validators) = config_payload.max_validators {
                if max_validators < min_validators {
                    return Err(
                        UpdateConfigError::MaxValidatorsCannotBeLowerThanMin,
                    );
                }
            };
            self.min_validators = min_validators;
        }
```

Reply:
Sure, but it’s easier to just unconditionally check afterwards.

Reply:
well i feel the above is much simpler, innit?

Reply:
Well, it has multiple nested `if`s and `if let Some(max_validators) = ...` is repeated twice so I’d argue is more complex.

Reply:
well even if u do outside, u would need a condition to get the value inside the option right?

Reply:
I’ve restructured the code. In the original when we were changing `self` it was enough to just verify values in `self` once they are changed. Now this can be done by unwrapping the old value.